Image uploader: Fix uploading of images to GCS

gastonqiu opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
Upload image to GCS with SDK to simplify the code. And make it easier to maintain.

**Which issue(s) this PR fixes**:
The GCS upload URL expires because the user can't upload the image file to GCS.

<!--

* Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #26194

**Special notes for your reviewer**:

Commit:
Chore: Fixed gcs uploader error due to url expired

Upload image to gcs with gcp sdk to simplify the code. And make it
more easy to maintain.

Commit:
chore: rename uploadFileWithSDK to uploadFile

aknuds1 commented on pkg/components/imguploader/gcsuploader.go:
```diff
@@ -43,40 +40,46 @@ func (u *GCSUploader) Upload(ctx context.Context, imageDiskPath string) (string,
 
-		u.log.Debug("Creating JWT conf")
-		conf, err := google.JWTConfigFromJSON(data, tokenUrl)
+		u.log.Debug("Creating google Credentials from json")
```
```suggestion
		u.log.Debug("Creating Google credentials from JSON")
```

aknuds1 commented on pkg/components/imguploader/gcsuploader.go:
```diff
@@ -43,40 +40,46 @@ func (u *GCSUploader) Upload(ctx context.Context, imageDiskPath string) (string,
-		u.log.Debug("Creating JWT conf")
-		conf, err := google.JWTConfigFromJSON(data, tokenUrl)
+		u.log.Debug("Creating google Credentials from json")
+		cred, err := google.CredentialsFromJSON(ctx, data)
```
```suggestion
		creds, err := google.CredentialsFromJSON(ctx, data)
```

aknuds1 commented on pkg/components/imguploader/gcsuploader.go:
```diff
@@ -43,40 +40,46 @@ func (u *GCSUploader) Upload(ctx context.Context, imageDiskPath string) (string,
 
-		u.log.Debug("Creating HTTP client")
-		client = conf.Client(ctx)
+		u.log.Debug("Creating gcs client")
```
```suggestion
		u.log.Debug("Creating GCS client")
```

aknuds1 commented on pkg/components/imguploader/gcsuploader.go:
```diff
@@ -43,40 +40,46 @@ func (u *GCSUploader) Upload(ctx context.Context, imageDiskPath string) (string,
 	} else {
-		u.log.Debug("Key file is empty, trying to use application default credentials")
-		client, err = google.DefaultClient(ctx)
+		u.log.Debug("Creating gcs client with no credentials")
```
```suggestion
		u.log.Debug("Creating GCS client without credentials")
```

aknuds1 commented on pkg/components/imguploader/gcsuploader.go:
```diff
@@ -85,26 +88,14 @@ func (u *GCSUploader) uploadFile(client *http.Client, imageDiskPath, key string)
-
-	req, err := http.NewRequest("POST", reqUrl, fileReader)
-	if err != nil {
+	u.log.Debug("Sending to gcs bucket using SDK")
```
```suggestion
	u.log.Debug("Sending to GCS bucket using SDK")
```

aknuds1 reviewed (commented):
I will have to test your PR before I can conclude on anything, but left some suggestions for improvements.

Commit:
Chore: Modify log messages

- Modify log message according to review #26493 to make it more clear.

aknuds1 commented:
Sorry, quite busy now, still have it on my agenda to test/review this.

gastonqiu commented:
It's OK. Take your time.

gastonqiu commented:
@aknuds1 This has been marked as stale. Can you help to remove the tag? Also does it help if I test this by myself and post the result here?

aknuds1 commented:
The stale bot applies the label automatically. I haven't forgot about the PR, but I'm very busy with a number of different PRs plus own tasks to take care of.

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into chore/gcsUploadErr

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into chore/gcsUploadErr

aknuds1 committed:
GCSUploader: Fix making of signed URL

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented on pkg/components/imguploader/gcsuploader.go:
```diff
@@ -65,76 +56,46 @@ func (u *GCSUploader) Upload(ctx context.Context, imageDiskPath string) (string,
-
-	u.log.Debug("Signing GCS URL")
-	var conf *jwt.Config
-	if u.keyFile != "" {
-		jsonKey, err := ioutil.ReadFile(u.keyFile)
-		if err != nil {
-			return "", fmt.Errorf("ioutil.ReadFile: %v", err)
-		}
-		conf, err = google.JWTConfigFromJSON(jsonKey)
-		if err != nil {
-			return "", fmt.Errorf("google.JWTConfigFromJSON: %v", err)
-		}
 	} else {
-		creds, err := google.FindDefaultCredentials(ctx, storage.ScopeReadWrite)
-		if err != nil {
-			return "", fmt.Errorf("google.FindDefaultCredentials: %v", err)
-		}
-		conf, err = google.JWTConfigFromJSON(creds.JSON)
+		u.log.Debug("Creating GCS client with default application credentials")
+		client, err = storage.NewClient(ctx)
 		if err != nil {
-			return "", fmt.Errorf("google.JWTConfigFromJSON: %v", err)
+			return "", err
 		}
 	}
-	opts := &storage.SignedURLOptions{
-		Scheme:         storage.SigningSchemeV4,
-		Method:         "GET",
-		GoogleAccessID: conf.Email,
-		PrivateKey:     conf.PrivateKey,
-		Expires:        time.Now().Add(u.signedUrlExpiration),
-	}
-	signedUrl, err := storage.SignedURL(u.bucket, key, opts)
+
+	err = u.uploadFile(ctx, client, imageDiskPath, key)
 	if err != nil {
-		return "", fmt.Errorf("storage.SignedURL: %v", err)
+		return "", err
 	}
-	return signedUrl, nil
```
Can you please tell me why this section, i.e. support for signed URLs, is removed?

gastonqiu replied:
You're right. It should not be removed since this is the new feature after I modify it. I don't notice when I rebase. I will add this back.

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into chore/gcsUploadErr

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into chore/gcsUploadErr

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into chore/gcsUploadErr

Commit:
Fix: The signed url should not be removed

- The signed url function had been removed accidentally after rebase
to master so I add the function back.

https://github.com/grafana/grafana/pull/26493#discussion_r503907199

Signed-off-by: gaston.qiu <gaston.qiu@umbocv.com>

Commit:
Fix: The signed url should not be removed

- The signed url function had been removed accidentally after rebase
to master so I add the function back.

https://github.com/grafana/grafana/pull/26493#discussion_r503907199

Signed-off-by: gaston.qiu <gaston.qiu@umbocv.com>

aknuds1 committed:
Merge branch 'chore/gcsUploadErr' of github.com:gastonqiu/grafana into chore/gcsUploadErr

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented on pkg/components/imguploader/gcsuploader.go:
```diff
@@ -65,35 +57,34 @@ func (u *GCSUploader) Upload(ctx context.Context, imageDiskPath string) (string,
-	if err != nil {
-		return "", err
-	}
+	err = u.uploadFile(ctx, client, imageDiskPath, key)
```
You need to keep the error check.
```suggestion
	err = u.uploadFile(ctx, client, imageDiskPath, key)
	if err != nil {
		return "", err
	}
```

Commit:
Fix: Return error when upload files failed

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into chore/gcsUploadErr

aknuds1 committed:
Merge branch 'chore/gcsUploadErr' of github.com:gastonqiu/grafana into chore/gcsUploadErr

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into chore/gcsUploadErr

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into chore/gcsUploadErr

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into chore/gcsUploadErr

aknuds1 commented:
Sorry for the long wait, but I'm testing the PR now.

aknuds1 committed:
Fix uploading to GCS

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Upgrade Go Storage SDK

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 reviewed (approved):
LGTM! Tested and works well both with and without signed URLs.

Please note that I had to rewrite a bit to make it function.

aknuds1 committed:
Clean up

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Clean up

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented:
Thanks for contributing!
